include/pybind11/factory.h

```diff
@@ -0,0 +1,316 @@
+    // must have an `Alias(Base &&)` constructor so that we can construct the alias from the base
+    // when needed.
+    template <typename Class, enable_if_t<Class::has_alias, int> = 0>
+    static void construct(Inst<Class> *self, Cpp<Class> &&result, PyTypeObject *cl_type) {
+        static_assert(std::is_move_constructible<Cpp<Class>>::value,
+            "pybind11::init() return-by-value factory function requires a movable class");
+        if (Py_TYPE(self) != cl_type) {
+            if (alias_constructible_from_cpp<Class>::value)
+                construct_alias_from_cpp<Class>(self, std::move(result));
+            else
+                throw type_error("pybind11::init(): unable to convert returned instance to "
+                                 "required alias class: no `Alias(Class &&)` constructor available");
+        }
+        else
+            new (self->value) Cpp<Class>(std::move(result));
+    }
+
+    // return-by-value version 3: the alias type itself--always initialize via the alias type (this
+    // is the factory equivalent of py::init_alias<...>()).
+    template <typename Class>
+    static void construct(Inst<Class> *self, Alias<Class> &&result, PyTypeObject *) {
+        static_assert(std::is_move_constructible<Alias<Class>>::value,
+            "pybind11::init() return-by-alias-value factory function requires a movable alias class");
+        new (self->value) Alias<Class>(std::move(result));
+    }
+
+    CFuncType class_factory;
+    AFuncType alias_factory;
+};
+
+// Helper definition to infer the detail::init_factory template type from a callable object
+template <typename Func, typename Return, typename... Args>
+init_factory<Func, Return, void, void, Args...> init_factory_decltype(Return (*)(Args...));
+template <typename Return1, typename Return2, typename... Args1, typename... Args2>
+inline constexpr bool init_factory_require_matching_arguments(Return1 (*)(Args1...), Return2 (*)(Args2...)) {
+    static_assert(sizeof...(Args1) == sizeof...(Args2),
+        "pybind11::init(class_factory, alias_factory): class and alias factories must have identical argument signatures");
+    static_assert(all_of<std::is_same<Args1, Args2>...>::value,
+        "pybind11::init(class_factory, alias_factory): class and alias factories must have identical argument signatures");
+    return true;
+}
+template <typename CFunc, typename AFunc,
+          typename CReturn, typename... CArgs, typename AReturn, typename... AArgs,
+          bool = init_factory_require_matching_arguments((CReturn (*)(CArgs...)) nullptr, (AReturn (*)(AArgs...)) nullptr)>
+init_factory<CFunc, CReturn, AFunc, AReturn, CArgs...> init_factory_decltype(
+    CReturn (*)(CArgs...), AReturn (*)(AArgs...));
+
+template <typename... Func> using init_factory_t = decltype(init_factory_decltype<Func...>(
+    (typename detail::remove_class<decltype(&std::remove_reference<Func>::type::operator())>::type *) nullptr...));
```

Comment:
When I tried this out with a plain function it didn't work. I suspect it's because a plain function doesn't have a `::operator()` when it tries to expand this. I can't think of a simple way to get the return type of either a function or a function object (`result_of` requires you to know the calling argument types). I guess it'll just need another traits helper class to handle the various cases.